src/api/plugin/plugin.ts

```diff
@@ -0,0 +1,99 @@
+import { Base } from '../base';
+import Transport from '../../transport/transport';
+import { notImplementedEnvErrorMsg } from '../../environment/environment';
+
+export interface PluginBare {
+    name: string;
+    version: string;
+}
+
+/**
+ * The Plugin API allows importing OpenFin plugins
+ * @namespace
+*/
+export default class Plugin extends Base {
+    private idBase: string;
+    private pluginsImportBaseKey: string; // unique plugins key for window
+    private noEsmSupportErrorMsg: string;
+    private importId: number;
+
+    constructor(wire: Transport) {
+        super(wire);
+
+        this.importId = 0;
+        this.idBase = wire.environment.getRandomId();
+        this.pluginsImportBaseKey = `__plugins_${this.idBase}_import_`;
+        this.noEsmSupportErrorMsg = 'ES modules are not supported in this version of OpenFin.';
+    }
+
+    /**
+     * Imports an OpenFin plugin and resolves with plugin's exported API.
+     *
+     * Works only in __OpenFin environment__.
+     *
+     * **Important**: If you set HTTP Content-Security-Policy's `script-src` directive
+     * you must allow `unsafe-inline` for `blob:` for this API to work.
+     *
+     * @param {Object} plugin - Plugin to import. Specified plugin must be listed in app's manifest.
+     * @param {string} plugin.name - plugin name
+     * @param {string} plugin.version - plugin version
+     * @return {Promise<any>}
+     * @tutorial Plugin.import
+     */
+    public async import(plugin: PluginBare): Promise<any> {
+        if (!this.isOpenFinEnvironment()) {
+            throw new Error(notImplementedEnvErrorMsg);
+        }
+
+        if (!this.isEsmSupported()) {
+            throw new Error(this.noEsmSupportErrorMsg);
+        }
+
+        const { payload } = await this.wire.sendAction('get-plugin-module', { plugin });
+        const { data: { _content } } = payload;
+
+        return this.importModule(_content);
+    }
+
+    // ESM is supported in OF v9+
+    private isEsmSupported = (): boolean => {
+        return parseInt((<any>window).fin.desktop.getVersion()) >= 9;
+    }
+
+    // Returns an unique key string for imports
+    private getImportKey = (): string => {
+        this.importId += 1;
+        return `${this.pluginsImportBaseKey}${this.importId}`;
+    }
+
+    // Handles importing a plugin. Resolves with a plugin API object.
+    private importModule(content: string): Promise<any> {
+        return new Promise((resolve, reject) => {
+            const blob = new Blob([content], { type: 'text/javascript' });
+            const url = window.URL.createObjectURL(blob);
```

Comment:
:neckbeard: can we delegate any OpenFin specific code to the environment class ?

Reply:
2 things:
• to look something like this `const url = this.wire.environment.createObjectUrl(blob)`?
• there is an environment check on line 44, does it change this point?